storage/boltdb/client_test.go

```diff
@@ -140,3 +141,189 @@ func BenchmarkSuiteLong(b *testing.B) {
 	}
 	testsuite.BenchmarkPathOperationsInLargeDb(b, longStore)
 }
+
+func BenchmarkClientWrite(b *testing.B) {
+	// setup db
+	ctx := testcontext.New(b)
+	defer ctx.Cleanup()
+	dbfile := ctx.File("testbolt.db")
+	dbs, err := NewShared(dbfile, "kbuckets", "nodes")
+	if err != nil {
+		b.Fatalf("failed to create db: %v\n", err)
+	}
+	defer func() {
+		if err := dbs[0].Close(); err != nil {
+			b.Fatalf("failed to close db: %v\n", err)
+		}
+		if err := dbs[1].Close(); err != nil {
+			b.Fatalf("failed to close db: %v\n", err)
+		}
+	}()
+	kdb := dbs[0]
+
+	// benchmark test: execute 1000 Put operations where each call to `PutAndCommit` does the following:
+	// 1) create a BoltDB transaction (tx), 2) execute the db operation, 3) commit the tx which writes it to disk.
+	for n := 0; n < b.N; n++ {
+		var wg sync.WaitGroup
+		for i := 0; i < 1000; i++ {
+			key := storage.Key(fmt.Sprintf("testkey%d", i))
+			value := storage.Value("testvalue")
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				err := kdb.PutAndCommit(key, value)
+				if err != nil {
+					b.Fatal("Put err:", err)
+				}
+			}()
+		}
+		wg.Wait()
+	}
+}
+
+func BenchmarkClientNoSyncWrite(b *testing.B) {
+	// setup db
+	ctx := testcontext.New(b)
+	defer ctx.Cleanup()
+	dbfile := ctx.File("testbolt.db")
+	dbs, err := NewShared(dbfile, "kbuckets", "nodes")
+	if err != nil {
+		b.Fatalf("failed to create db: %v\n", err)
+	}
+	defer func() {
+		if err := dbs[0].Close(); err != nil {
+			b.Fatalf("failed to close db: %v\n", err)
+		}
+		if err := dbs[1].Close(); err != nil {
+			b.Fatalf("failed to close db: %v\n", err)
+		}
+	}()
+	kdb := dbs[0]
+
+	// benchmark test: execute 1000 Put operations with fsync turned off.
+	// Each call to `PutAndCommit` does the following: 1) creates a BoltDB transaction (tx),
+	// 2) executes the db operation, and 3) commits the tx which does NOT write it to disk.
+	kdb.db.NoSync = true
+	for n := 0; n < b.N; n++ {
+		var wg sync.WaitGroup
+		for i := 0; i < 1000; i++ {
+			key := storage.Key(fmt.Sprintf("testkey%d", i))
+			value := storage.Value("testvalue")
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				err := kdb.PutAndCommit(key, value)
+				if err != nil {
+					b.Fatal("PutAndCommit Nosync err:", err)
```

Comment:
benchmark methods cannot be called from a goroutine it's a race.

Reply:
oops I will fix that in another PR since this is merged. thanks egon.